### PR TITLE
[action] add `xcodes` action, deprecating `xcversion` and `xcode-install`

### PR DESCRIPTION
**Pull request by rogerluan**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

With the planned sunset of `xcode-install` (aka `xcversion`) gem (https://github.com/xcpretty/xcode-install/issues/467), a new action was needed to support `xcodes`.

### Description

This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 and https://github.com/RobotsAndPencils/xcodes/pull/182 merge & release so the `installed` option with specified version is available, as well as `--install` and `--select` arguments for the `install` option.

These PRs were later merged & released under v1.1.0 https://github.com/RobotsAndPencils/xcodes/releases/tag/1.1.0, so v1.1.0 of xcodes is required to use and test this feature.

### Testing Steps

To test this branch, modify your Gemfile as:

```ruby
gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'rogerluan-xcodes'
```

And run `bundle install` to apply the changes. 

Then switch your implementation of `xcversion` and `xcode_install` according to the migration guide which can be found here: https://github.com/xcpretty/xcode-install/blob/master/MIGRATION.md

Any and every testing is highly appreciated! I did very limited testing only 🙇 

image

image

### CODE REVIEW

**Comment by rogerluan** [comment]
The spec failure that is happening in this PR also happens in master, and it may be something related to recent changes on Apple's end 😬

**Comment by rogerluan** [comment]
It's weird:
- Locally, on my machine, a lot of specs fail both in this PR as well as in master, same for rubocop violations
- In CI, master seems to not be failing
- In CI, in another PR of mine, all specs passed
- In CI, this PR keeps failing in non-related changes 🤔

**Comment by rogerluan** [comment]
Ugh, so, locally, here's the breakdown:

- Ruby 3.1.x: a bunch of linter and specs issues
- Ruby 3.x.x: linter and specs are all passing
- Ruby 2.7.x: linter and specs are all passing
- Ruby 2.6.x: I'm having trouble installing this version using rbenv (M2 machine) 😥 so I couldn't test..

CI is failing specs and it is running Ruby 2.6.0 apparently... but not sure why this is happening in my PR only, and on presumably unrelated files (`fastlane_core/spec/cert_checker_spec.rb`) something related to:

```
#installed_wwdr_certificates
      should return installed certificate's alias (FAILED - 4)
      should return an empty array if unknown WWDR certificates are found (FAILED - 5)
```

😬 Any help would be appreciated

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (revolter) [inline fastlane/lib/fastlane/actions/xcodes.rb:118]:
```suggestion
          ['XCODES_XCODE_PATH', 'The path to the newly installed Xcode version']
```

for consistency with the `return_value`.

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (revolter) [inline fastlane/lib/fastlane/actions/xcodes.rb:15]: What would be the issue if we allow updating the list even if we want to select it for the current build only?

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (revolter) [inline fastlane/lib/fastlane/actions/xcodes.rb:41]: This doesn't work (anymore?):

```plaintext
[21:10:54]: --------------------
[21:10:54]: --- Step: xcodes ---
[21:10:54]: --------------------
[21:10:54]: $ /opt/homebrew/bin/xcodes installed '13.3'
[21:10:54]: ▸ Error: Unexpected argument '13.3'
[21:10:54]: ▸ Usage: xcodes installed [--directory <directory>] [--color] [--no-color]
[21:10:54]: ▸ See 'xcodes installed --help' for more information.
+------------------+------------------+
|            Lane Context             |
+------------------+------------------+
| DEFAULT_PLATFORM | ios              |
| PLATFORM_NAME    | ios              |
| LANE_NAME        | ios ensure_xcode |
+------------------+------------------+
[21:10:54]: Exit status of command '/opt/homebrew/bin/xcodes installed '13.3'' was 64 instead of 0.
Error: Unexpected argument '13.3'
Usage: xcodes installed [--directory <directory>] [--color] [--no-color]
  See 'xcodes installed --help' for more information.
```

I'm using `xcodes` version `1.0.0`.

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:2]: This is more of a curiosity, but, what is the role of `#`?

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:30]:
```suggestion
      expect(subject.class).to receive(:find_xcodes_binary_path).and_return("/path/to/bin/xcodes")
      expect(subject.class.find_xcodes_binary_path).to eq("/path/to/bin/xcodes")
```

for consistency with [xcodes_spec.rb](https://github.com/fastlane/fastlane/pull/20672/files#diff-2ca850ff6a7654f270c5deb87cbb737659ee2ee968e955924ae6524b170b78b5R6)?

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:44]:
```suggestion
        it "raises user error" do
```

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:45]:
```suggestion
          expect { subject.class::Verify.requirement(argument) }.to raise_error(FastlaneCore::Interface::FastlaneError, /Version must be specified/)
```

Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:45]:
```suggestion
          expect { subject.class::Verify.requirement(nil) }.to raise_error(FastlaneCore::Interface::FastlaneError, 'Version must be specified')
```

? Or am I missing something? 🤔

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:52]:
```suggestion
        it "raises user error" do
```

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (revolter) [inline fastlane/spec/helper/xcodes_helper_spec.rb:21]:
```suggestion
      it "returns nil" do
```

**Review by revolter — CHANGES_REQUESTED** [review]
Hi @rogerluan 👋🏻

Nice to "see" you again after a long time 😃 And sorry for the late reply.

---

> 😬 Any help would be appreciated

I can't promise anything, but I hope I'll take a look some time after reviewing this.

---

I read the migration guide and this PR's description, I started testing it, but it failed with:

```plaintext
[21:00:46]: Couldn't find xcodes binary at path ''
[21:00:46]: You passed invalid parameters to 'xcodes'.
[21:00:46]: Check out the error below and available options by running `fastlane action xcodes`
+------------------+------------------+
|            Lane Context             |
+------------------+------------------+
| DEFAULT_PLATFORM | ios              |
| PLATFORM_NAME    | ios              |
| LANE_NAME        | ios ensure_xcode |
+------------------+------------------+
[21:00:46]: Invalid default value for binary_path, doesn't match verify_block
```

Although it is fairly obvious that the issue is that it can't find `xcodes`, and you have to (most probably) run `brew install robotsandpencils/made/xcodes`, I think that this should be mentioned in the migration guide and/or the new action's description, since it's a required dependency that's not automatically managed by _fastlane_. What do you think?

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:41]: It doesn't work 'yet' 🙈 This PR this currently blocked by the release of the next version of `xcodes` 😬 as I stated in the PR description:

>This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 merge & release so the installed option with specified version is available.

It has been merged, but not yet released. I've contacted the maintainer of xcodes (@MattKiazyk 🙌 ) and he said he doesn't have a release cadence, so yeah we need to wait for the next release to see this feature available

Any suggestions to enhance this @revolter ? 🙇

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:15]: Prior to release v1.0.0, updating the list required Apple authentication, so since update_list is true by default, if you just wanted to pass `select_for_current_build_only: true`, you wouldn't want to be bothered to authenticate with Apple. Now it doesn't require authentication anymore, so I don't see any issues with that!

@MattKiazyk could you confirm what is the exact behavior of the new auth-less requests in v1.0.0? Is it true that there are zero chances of requiring an authentication, regardless of the flow, regardless of which Xcode versions are already installed, macOS version, etc?

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb:2]: >use the Ruby documentation convention of . (or ::) when referring to a class method's name and # when referring to an instance method's name.

Source: https://www.betterspecs.org

(I learn these things as I go 😆 Ruby is not my mother tongue, clearly haha)

Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb:2]: 🤔 Which made me think... `read_xcode_version_file` is actually a class method, not an instance one. I should probably fix this lol

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb:30]: Oops, good catch! Left this behind by mistake after doing some local tests 😬

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb:45]: Nope, you're right :)

Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb:45]:
```suggestion
          expect { subject.class::Verify.requirement(argument) }.to raise_error(FastlaneCore::Interface::FastlaneError, 'Version must be specified')
```

File: fastlane/spec/helper/xcodes_helper_spec.rb
Comment (rogerluan) [reply fastlane/spec/helper/xcodes_helper_spec.rb]: Done :)

**Comment by rogerluan** [comment]
Oops, forgot to address the "no binary" issue. Gimme a few mins

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:15]: If no user/pass is required EVER ☝️ then I can go ahead and remove all the user/pass management completely from this new action ✨

**Comment by rogerluan** [comment]
Done @revolter 😊 Nice catch!

image

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (revolter) [reply fastlane/lib/fastlane/actions/xcodes.rb:41]: > This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 merge & release so the installed option with specified version is available.

Oh, I missed the "release" part 🤦🏻‍♂️ 😂

> Any suggestions to enhance this @revolter?

Enhance what?

Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:41]: @revolter nevermind, I thought of adding a version check but there wouldn't be a great way to implement the same functionality before that feature gets released on `xcodes`. Let's just wait 🙇

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (MattKiazyk) [reply fastlane/lib/fastlane/actions/xcodes.rb:15]: @rogerluan  On Xcodes - if using the xcoderelease.com api (default) it is not required on the latest version. If using the Apple source it still would be.

Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:15]: Thanks @MattKiazyk ! 🙏

**Comment by rogerluan** [comment]
@revolter would you mind helping me understand what's failing in the CI specs? I added comments about it [here](https://github.com/fastlane/fastlane/pull/20672#issuecomment-1252236081) and I'm still clueless on how to solve this 😿

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb]: I just removed the requirement to provide xcodes action Apple user/pass 💪

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (rogerluan) [reply fastlane/lib/fastlane/actions/xcodes.rb:34]: I ended up adding this version check (for v1.1.0) just for validation sake (it crashes if <1.1.0) so users are aware of this limitation 👍

File: fastlane/lib/fastlane/actions/xcodes.rb
Comment (revolter) [inline fastlane/lib/fastlane/actions/xcodes.rb:22]: Inline this `shell_command` variable, and also the one from line `36`?

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (revolter) [inline fastlane/spec/actions_specs/xcodes_spec.rb:25]:
```suggestion
          expect(Fastlane::Actions).to receive(:sh).with('--update')
```

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (revolter) [inline fastlane/spec/actions_specs/xcodes_spec.rb:33]:
```suggestion
          expect(Fastlane::Actions).to_not receive(:sh).with('--update')
```

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (revolter) [inline fastlane/spec/actions_specs/xcodes_spec.rb:42]:
```suggestion
        expect(Fastlane::Actions).to receive(:sh).with('installed')
```

**Review by revolter — CHANGES_REQUESTED** [review]
Just some small changes, to be less confusing for future contributors 🤔

**Comment by revolter** [comment]
@rogerluan, would you mind if I'd remove the merge commit, and rebase the branch instead?

**Comment by revolter** [comment]
> Would that help debug the CI problem?

I hope it fixes it 😂 Because it doesn't make sense

**Comment by revolter** [comment]
It doesn't :( I'll test some ideas this evening.

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (rogerluan) [reply fastlane/spec/actions_specs/xcodes_spec.rb:25]: This is intentionally a regex, rspec treat string literals and regexes differently:

image

If I were to use a string, I'd not only have to use the entire exact string, but also expect the shell commands before and after the `/--update/` one 🤷 

This applies to all of the ones you commented on here

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (rogerluan) [reply fastlane/spec/actions_specs/xcodes_spec.rb]: See above

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (rogerluan) [reply fastlane/spec/actions_specs/xcodes_spec.rb:42]: See above

File: fastlane/spec/actions_specs/xcodes_spec.rb
Comment (revolter) [reply fastlane/spec/actions_specs/xcodes_spec.rb:25]: True, my bad 🤦🏻‍♂️

**Comment by revolter** [comment]
Ok, this took me a lot longer than I thought it would :| Very very weird issue, but great debugging nonetheless ❤️

I also noticed that some paths were being concatenated using `+`, so I used the same method as in `xcode_select.rb` (note: it works with an already present slash because that method makes sure not to duplicate the separator: https://stackoverflow.com/a/4114966/865175).

There are still some failures, but these ones might not be that hard to fix 🤔

_(Also, because I rebased and force-pushed to the branch, guessing that you didn't commit anything after 136b6648, please delete the local branch, fetch the new changes, and check it out again, to avoid other superfluous merges.)_

I really need to go to sleep now, as it's already tomorrow 😫

**Comment by revolter** [comment]
Please let me work on this more? 😂 I have an idea that I want to test this evening.

**Comment by revolter** [comment]
Ok, I was not happy with the fix, because I didn't understand it fully (still don't), and felt more like a hack. From what I (extensively) read, even through [the source code](https://github.com/ruby/open3/blob/master/lib/open3.rb), and by also testing it locally by replacing the `curl` command with `ls`, it looks like this (the last commit) is the safe and correct way to fix this issue for good.

**Comment by revolter** [comment]
Somehow, the 2 failures (Validate Documentation and Validate Fastlane.swift generation) got fixed too 😮

**Review by revolter — APPROVED** [review]

File: fastlane_core/lib/fastlane_core/cert_checker.rb
Comment (rogerluan) [inline fastlane_core/lib/fastlane_core/cert_checker.rb:169]: 👏👏👏

File: fastlane/lib/fastlane/helper/xcodes_helper.rb
Comment (rogerluan) [inline fastlane/lib/fastlane/helper/xcodes_helper.rb]: What the... 🤯

**Comment by rogerluan** [comment]
@revolter dude, this was just amazing!! Thank you thank you thank you ❤️ Great context that you provided in the commit messages also! I'd never have guessed something like this, I can't even imagine how was your train of thought and how you got to the root cause 🤯 

Separately, do you think we should at least create an Issue for future improvement of getting rid of all the `$?` in other specs (and maybe in the source code too)?

Again, thank you so much 🙇 I'm so glad we'll be able to move this PR forward! 💪

**Review by rogerluan — COMMENTED** [review]
@revolter dude, this was just amazing!! Thank you thank you thank you ❤️ Great context that you provided in the commit messages also! I'd never have guessed something like this, I can't even imagine how was your train of thought and how you got to the root cause 🤯 

Separately, do you think we should at least create an Issue for future improvement of getting rid of all the `$?` in other specs (and maybe in the source code too)?

Again, thank you so much 🙇 I'm so glad we'll be able to deliver this PR to `xcode-install` users  💪

**Comment by revolter** [comment]
> I can't even imagine how was your train of thought and how you got to the root cause 🤯

Brute-force trial and error, using divide et impera, by reverting half of the changes, then half of the half, etc., until I pinpointed the responsible line of code, and then a lightbulb went off in my head about the `$?` usage from `cert_checker.rb`.

image

> Separately, do you think we should at least create an Issue for future improvement of getting rid of all the $? in other specs (and maybe in the source code too)?

Hmm, maybe that would be the way to go, but it's also a case of "if it ain't broken, don't fix it" 🤔 Maybe the best thing is that at least the 2 of us know of this, and hopefully remember it when (if) it happens in the future.

> Again, thank you so much 🙇 I'm so glad we'll be able to deliver this PR to `xcode-install` users  💪 

Really happy that I could (finally) help you with this one. Thank you too for the kind words, I'm really flattered ☺️ I'm happy that you took the time to do the other 99% of work for this migration ❤️

**Review by joshdholtz — APPROVED** [review]
This is soooo amazing!! ❤️ Sorry it took so long but I'm back from family leave and thank for doing this 😊

**Pull request merged**
